magma/clock.py

```diff
@@ -134,29 +134,51 @@ def first(iterable):
     return next((item for item in iterable if item is not None), None)
 
 
-def get_first_clock(port, clocktype):
+def get_clocks(port, clocktype):
+    if port.is_input():
+        # We are looking for default drivers (outputs)
+        return []
     if isinstance(port, clocktype):
-        return port
+        return [port]
     if isinstance(port, Tuple):
-        clks = (get_first_clock(elem, clocktype) for elem in port)
-        return first(clks)
+        clks = []
+        for elem in port:
+            clks += get_clocks(elem, clocktype)
+        return clks
     if isinstance(port, Array):
-        return get_first_clock(port[0], clocktype)
-    return None
+        clks = get_clocks(port[0], clocktype)
+        if not clks:
+            # early exit to avoid traversing children when unnecessary
+            return clks
+        for elem in port[1:]:
+            clks += get_clocks(elem, clocktype)
+        return clks
+    return []
+
+
+def get_clocks_from_defn(defn, clocktype):
```

Comment:
I think we should make this return a generator rather than a list for performance, since we only every get the first element. I think we can use `yield` + `itertools.chain` to do this concisely. Happy to help make this change.